src/idl/src/tree.c

```diff
@@ -983,6 +984,202 @@ static const char *describe_struct(const void *ptr)
   return "struct";
 }
 
+static int compare_declarator(const void *lhs, const void *rhs)
+{
+  const idl_declarator_t *d_lhs = *(const idl_declarator_t **)lhs,
+                         *d_rhs = *(const idl_declarator_t **)rhs;
+  if (d_lhs->id.value < d_rhs->id.value)
+    return -1;
+  else if (d_lhs->id.value > d_rhs->id.value)
+    return 1;
+  else
+    return 0;
+}
+
+idl_retcode_t
+idl_propagate_autoid(
+  idl_pstate_t *pstate,
+  idl_node_t *list,
+  idl_autoid_t toset)
+{
+  assert(list);
+
+  idl_node_t *node = NULL;
+  idl_retcode_t ret = IDL_RETCODE_OK;
+  IDL_FOREACH(node, list) {
+    if (idl_is_module(node)) {
+      idl_module_t *mod = (idl_module_t*)node;
+      if (!mod->autoid.annotation) {
+        mod->autoid.value = toset;
+        ret = idl_propagate_autoid(pstate, mod->definitions, toset);
+      }
+    } else if (idl_is_struct(node)) {
+      idl_struct_t *str = (idl_struct_t*)node;
+      if (!str->autoid.annotation) {
+        str->autoid.value = toset;
+        ret = idl_propagate_autoid_struct(pstate, str);
+      }
+    } else if (idl_is_union(node)) {
+      idl_union_t *u = (idl_union_t*)node;
+      if (!u->autoid.annotation) {
+        u->autoid.value = toset;
+        ret = idl_propagate_autoid_union(pstate, u);
+      }
+    }
+    if (ret)
+      break;
+  }
+
+  return ret;
+}
+
+static void
+process_declarator_id(
+  idl_declarator_t *decl,
+  idl_declarator_t *last,
+  idl_autoid_t autoid)
+{
+  assert(decl);
+
+  if (decl->id.annotation) {
+    /*this declarator already has an ID annotation*/
+  } else if (autoid == IDL_HASH) {
+    assert(decl->name->identifier);
+    decl->id.value = idl_hashid(decl->name->identifier);
+  } else if (last) {
+    decl->id.value = (last->id.value + 1) & 0x0FFFFFFFu;
+  } else {
+    decl->id.value = 0;
+  }
+}
+
+idl_retcode_t
+idl_propagate_autoid_struct(
+  idl_pstate_t *pstate,
+  idl_struct_t *str)
+{
+  assert(str);
+
+  size_t n_ids = 0;
+  idl_member_t *mem = NULL;
+  idl_declarator_t *decl = NULL, *strlast = NULL, *last = NULL;
+  idl_retcode_t ret = IDL_RETCODE_OK;
+
+  idl_struct_t *current = str;
+  while (current) {
+    IDL_FOREACH(mem, current->members) {
+      IDL_FOREACH(decl, mem->declarators) {
+        n_ids++;
+        if (current != str && !last)
+          strlast = decl;
+      }
+    }
+    if (!last)
+      last = strlast;
```

Comment:
To work around inheritance for iteration wrt struct vs union, we can just always set last like with union traversal and then check if we shouldn't set it to `NULL` after if the ancestor of the declarator is the same as the node we started at.

Reply:
Made a more generic implementation, so this is no longer applicable.